src/client.rs

```diff
@@ -75,48 +86,65 @@ impl Client {
 
         let mut buckets_in_region: Vec<BucketItem> = Vec::new();
         for bucket in buckets {
-            let bucket_location = self
-                .client
-                .get_bucket_location()
-                .bucket(bucket.name.as_str())
-                .send()
-                .await;
+            let region = self.get_bucket_region(&bucket.name).await.unwrap();
+            if region == self.region {
+                buckets_in_region.push(bucket);
+            }
+        }
 
-            let location = bucket_location.map_err(|e| {
+        self.bucket_region_cache.write_cache().unwrap();
+
+        Ok(buckets_in_region)
+    }
+
+    async fn get_bucket_region(&self, bucket_name: &str) -> Result<String> {
+        if let Some(location) = self.bucket_region_cache.get(bucket_name) {
+            return Ok(location);
+        }
+
+        let bucket_region = self
+            .client
+            .get_bucket_location()
+            .bucket(bucket_name)
+            .send()
+            .await
+            .map_err(|e| {
                 AppError::new(
-                    format!("Failed to get bucket location for bucket {}", bucket.name),
+                    format!("Failed to fetch region for bucket {}", bucket_name),
                     e,
                 )
-            })?;
-
-            if let Some(constraint) = location.location_constraint() {
-                let constraint_str = match constraint.as_str() {
-                    // For us-east-1, `location.location_constraint()` returns Some<Unknown<UnknownVariantValue("")>>, with it's `as_str()` returning "".
-                    // We explicitly handle this case by mapping it to "us-east-1".
-                    "" => "us-east-1",
-                    other => other,
-                };
-
-                if constraint_str == self.region {
-                    buckets_in_region.push(bucket);
+            })?
+            .location_constraint()
+            .map(|loc| {
+                if loc.as_str().is_empty() {
+                    // Location.location_constraint() returns empty string for us-east-1 as it's the default region for S3.
+                    // Map it to us-east-1 to avoid having to deal with it as a special case.
+                    "us-east-1".to_string()
```

Comment:
As mentioned in the original comment, it's important to note that it returns `Some<Unknown<UnknownVariantValue("")>>` instead of `None`.

Reply:
Yes good catch. I forgot about the docs being bad here and the need for clarifying that part of it.